# CI7340 - ADP — Data Cleaning & Transformation - Kevin Gomez
**Dataset:** Montgomery County Crime (2018–2022)  
**Goal:** Inspect → Clean → Deduplicate → Add helper fields → Save clean dataset + brief reports

## 1. Setup and Configuration

**Purpose:** Import required libraries and define global configuration variables for the data cleaning pipeline.

**What this cell does:**
- Imports essential libraries: `pandas` for data manipulation, `numpy` for numerical operations, `pyarrow` for efficient parquet file handling
- Suppresses datetime parsing warnings for cleaner output
- Defines input/output paths for raw data, cleaned data, and reports
- Sets data quality parameters (date columns to parse, sparse columns to drop, year range)
- Configures pandas display options for better data inspection

**Key Configuration:**
- `YEAR_MIN, YEAR_MAX`: Filter data to 2018-2022 as per coursework scope
- `OPTIONALLY_DROP_IF_VERY_SPARSE`: Columns with >90% missing values will be removed


In [1]:
from __future__ import annotations
import io
from pathlib import Path
import numpy as np
import pandas as pd, pyarrow
from IPython.display import Markdown, display
import warnings

# Suppress datetime parsing warnings for cleaner output
warnings.filterwarnings("ignore", message="Could not infer format")

# ========== INPUT/OUTPUT PATHS ==========
# Raw crime dataset from Montgomery County (2018-2022)
RAW_CSV_PATH = "Crime_Dataset_Coursework1-1.csv"

# Create output directories for processed data and reports
OUT_DIR_DATA = Path("data/processed")
OUT_DIR_REPORTS = Path("reports")
OUT_DIR_FIGS = Path("reports/figs")

for d in [OUT_DIR_DATA, OUT_DIR_REPORTS, OUT_DIR_FIGS]:
    d.mkdir(parents=True, exist_ok=True)  # Create if doesn't exist

# Define output file paths
CLEAN_CSV_PATH = OUT_DIR_DATA / "crime_clean.csv"
CLEAN_PARQUET_PATH = OUT_DIR_DATA / "crime_clean.parquet"
MISSING_BEFORE_CSV = OUT_DIR_REPORTS / "missingness_before.csv"
MISSING_AFTER_CSV = OUT_DIR_REPORTS / "missingness_after.csv"
DQ_SUMMARY_MD = OUT_DIR_REPORTS / "data_quality_summary.md"
DATA_DICTIONARY_MD = OUT_DIR_REPORTS / "data_dictionary.md"

# ========== DATA QUALITY PARAMETERS ==========
# Columns containing date/time information to parse
DATE_COLUMNS = ["Dispatch Date / Time", "Start_Date_Time", "End_Date_Time"]

# Columns to drop if they have >90% missing values (very sparse data)
OPTIONALLY_DROP_IF_VERY_SPARSE = ["Street Prefix", "Street Suffix"]

# Year range filter (coursework scope: 2018-2022)
YEAR_MIN, YEAR_MAX = 2018, 2022

# ========== PANDAS DISPLAY OPTIONS ==========
# Configure pandas to show more rows/columns for inspection
pd.set_option("display.max_rows", 100)
pd.set_option("display.max_columns", 200)

print("✓ Libraries imported successfully")
print(f"✓ Output directories created: {OUT_DIR_DATA}, {OUT_DIR_REPORTS}, {OUT_DIR_FIGS}")
print(f"✓ Configuration set: Year range [{YEAR_MIN}-{YEAR_MAX}]")

✓ Libraries imported successfully
✓ Output directories created: data/processed, reports, reports/figs
✓ Configuration set: Year range [2018-2022]


## 2. Load Raw Dataset

**Purpose:** Load the raw crime dataset and perform initial inspection.

**What this cell does:**
- Reads the CSV file using `pd.read_csv()` with `low_memory=False` to prevent data type inconsistencies
- Displays dataset dimensions (rows × columns)
- Shows first few rows to understand data structure

**Expected Output:**
- Dataset shape (should be ~306,000 rows)
- Preview of crime incident records with all original columns


In [2]:
# Load raw crime dataset from CSV
# low_memory=False ensures consistent data type inference across chunks
df_raw = pd.read_csv(RAW_CSV_PATH, low_memory=False)

# Display dataset dimensions
print(f"Dataset Shape: {df_raw.shape[0]:,} rows × {df_raw.shape[1]} columns")
print(f"Total cells: {df_raw.size:,}")
print(f"Memory usage: {df_raw.memory_usage(deep=True).sum() / 1024**2:.2f} MB")

# Show first few rows for initial inspection
df_raw.head()

Dataset Shape: 306,094 rows × 30 columns
Total cells: 9,182,820
Memory usage: 446.18 MB


,Incident ID,Offence Code,CR Number,Dispatch Date / Time,NIBRS Code,Victims,Crime Name1,Crime Name2,Crime Name3,Police District Name,Block Address,City,State,Zip Code,Agency,Place,Sector,Beat,PRA,Address Number,Street Prefix,Street Name,Street Suffix,Street Type,Start_Date_Time,End_Date_Time,Latitude,Longitude,Police District Number,Location
0,201202980,3550,180042096,08/23/2018 09:52:08 PM,35B,1,Crime Against Society,Drug Equipment Violations,DRUGS - NARCOTIC EQUIP - POSSESS,GERMANTOWN,12800 BLK MIDDLEBROOK RD,GERMANTOWN,MD,20874.0,MCPD,Street - In vehicle,N,5N1,447,12800.0,NaN,MIDDLEBROOK,NaN,RD,08/23/2018 09:52:00 PM,NaN,39.177744,-77.265619,5D,"(39.1777, -77.2656)"
1,201181293,3522,180015424,03/30/2018 01:00:55 AM,35A,1,Crime Against Society,Drug/Narcotic Violations,DRUGS - OPIUM OR DERIVATIVE - POSSESS,BETHESDA,8300 BLK WOODMONT AVE,BETHESDA,MD,20814.0,MCPD,Street - In vehicle,E,2E2,054,8300.0,NaN,WOODMONT,NaN,AVE,03/30/2018 01:01:00 AM,NaN,38.992693,-77.097063,2D,"(38.9927, -77.0971)"
2,201181293,3562,180015424,03/30/2018 01:00:55 AM,35A,1,Crime Against Society,Drug/Narcotic Violations,DRUGS - MARIJUANA - POSSESS,BETHESDA,8300 BLK WOODMONT AVE,BETHESDA,MD,20814.0,MCPD,Street - In vehicle,E,2E2,054,8300.0,NaN,WOODMONT,NaN,AVE,03/30/2018 01:01:00 AM,NaN,38.992693,-77.097063,2D,"(38.9927, -77.0971)"
3,201193163,3520,180029476,06/14/2018 10:26:45 PM,35A,1,Crime Against Society,Drug/Narcotic Violations,DRUGS - OPIUM OR DERIVATIVE - SELL,MONTGOMERY VILLAGE,400 BLK QUINCE ORCHARD RD,GAITHERSBURG,MD,20878.0,GPD,Hotel/Motel/Etc.,P,6P2,445,400.0,NaN,QUINCE ORCHARD,NaN,RD,06/14/2018 10:26:00 PM,06/15/2018 03:00:00 AM,39.147954,-77.218189,6D,"(39.148, -77.2182)"
4,201204355,2204,180043926,09/03/2018 12:06:54 PM,220,1,Crime Against Property,Burglary/Breaking and Entering,BURGLARY - NO FORCED ENTRY-RESIDENTIAL,BETHESDA,4800 BLK FALSTONE AVE,CHEVY CHASE,MD,20815.0,MCPD,Residence - Single Family,E,2E1,017,4800.0,NaN,FALSTONE,NaN,AVE,09/02/2018 11:30:00 PM,09/03/2018 12:30:00 AM,38.966174,-77.096561,2D,"(38.9662, -77.0966)"


## 3. Initial Data Inspection

**Purpose:** Understand the dataset structure, data types, and column names.

**What this cell does:**
- Uses `df.info()` to display:
  - Column names and count
  - Data types for each column
  - Non-null counts (helps identify missing values)
  - Memory usage
- Lists all column names for reference

**Why this matters:**
- Identifies which columns need type conversion (e.g., dates stored as strings)
- Shows initial data completeness
- Helps plan data cleaning strategies


In [3]:
# Display detailed information about the dataset
print("=" * 80)
print("DATASET INFORMATION")
print("=" * 80)
df_raw.info()

print("\n" + "=" * 80)
print("COLUMN NAMES")
print("=" * 80)
# Display all column names for easy reference
columns_list = df_raw.columns.to_list()
for i, col in enumerate(columns_list, 1):
    print(f"{i:2d}. {col}")
    
print(f"\nTotal columns: {len(columns_list)}")

DATASET INFORMATION
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 306094 entries, 0 to 306093
Data columns (total 30 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   Incident ID             306094 non-null  int64  
 1   Offence Code            306094 non-null  object 
 2   CR Number               306094 non-null  int64  
 3   Dispatch Date / Time    257065 non-null  object 
 4   NIBRS Code              306094 non-null  object 
 5   Victims                 306094 non-null  int64  
 6   Crime Name1             305822 non-null  object 
 7   Crime Name2             305822 non-null  object 
 8   Crime Name3             305822 non-null  object 
 9   Police District Name    306000 non-null  object 
 10  Block Address           279888 non-null  object 
 11  City                    304818 non-null  object 
 12  State                   306094 non-null  object 
 13  Zip Code                302915 non-null  float64
 14  

## 4. Missing Value Analysis (BEFORE Cleaning)

**Purpose:** Quantify and analyze missing data patterns in the raw dataset.

**What this cell does:**
- Calculates missing value ratio for each column
- Converts to percentage for easier interpretation
- Sorts columns by missingness (highest to lowest)
- Saves results to CSV for documentation
- Displays top 15 columns with most missing values

**Why this matters:**
- Identifies data quality issues
- Informs imputation/deletion strategies
- Provides baseline for measuring cleaning effectiveness
- Documents initial data quality for the coursework report

**Expected findings:**
- Some columns (Street Prefix/Suffix) may have >90% missing
- Address and time fields may have moderate missingness


In [4]:
# Calculate missing value statistics for each column
miss_before = df_raw.isna().mean().sort_values(ascending=False).to_frame("missing_ratio")
miss_before["missing_pct"] = (miss_before["missing_ratio"] * 100).round(2)
miss_before["missing_count"] = df_raw.isna().sum().sort_values(ascending=False)

# Save to CSV for documentation
miss_before.to_csv(MISSING_BEFORE_CSV)
print(f"✓ Missing value analysis saved to: {MISSING_BEFORE_CSV}")

# Display summary statistics
print(f"\nTotal missing cells in dataset: {df_raw.isna().sum().sum():,}")
print(f"Percentage of dataset that is missing: {(df_raw.isna().sum().sum() / df_raw.size * 100):.2f}%")
print(f"\nColumns with >50% missing: {(miss_before['missing_pct'] > 50).sum()}")
print(f"Columns with >90% missing: {(miss_before['missing_pct'] > 90).sum()}")

print("\n" + "=" * 80)
print("TOP 15 COLUMNS BY MISSINGNESS")
print("=" * 80)
miss_before.head(15)

✓ Missing value analysis saved to: reports/missingness_before.csv

Total missing cells in dataset: 865,131
Percentage of dataset that is missing: 9.42%

Columns with >50% missing: 3
Columns with >90% missing: 2

TOP 15 COLUMNS BY MISSINGNESS


,missing_ratio,missing_pct,missing_count
Street Suffix,0.982254,98.23,300662
Street Prefix,0.955468,95.55,292463
End_Date_Time,0.528132,52.81,161658
Dispatch Date / Time,0.160176,16.02,49029
Block Address,0.085614,8.56,26206
Address Number,0.085297,8.53,26109
Zip Code,0.010386,1.04,3179
Sector,0.004998,0.50,1530
Beat,0.004998,0.50,1530
City,0.004169,0.42,1276


## 5. Define Helper Functions for Data Cleaning

**Purpose:** Create reusable utility functions for consistent data transformation.

**Functions defined:**

1. **`normalise_text_series(s)`**
   - Strips leading/trailing whitespace
   - Collapses multiple spaces into single space
   - Converts to title case (e.g., "SILVER SPRING" → "Silver Spring")
   - Handles missing values appropriately
   - Used for: crime names, locations, agencies, street names

2. **`to_datetime_safe(s)`**
   - Converts strings to datetime objects
   - Uses `errors='coerce'` to convert invalid dates to NaT instead of raising errors
   - Used for: all date/time columns

3. **`coerce_zip(zip_series)`**
   - Converts to string format
   - Removes non-numeric characters
   - Pads to 5 digits (e.g., 123 → 00123)
   - Handles floats and strings consistently
   - Replaces empty strings with NA

4. **`choose_primary_offense(row)`**
   - Extracts the first non-null value from Crime Name1, Crime Name2, Crime Name3
   - Creates a single "Primary Offense" field
   - Simplifies analysis by consolidating multiple offense columns

**Why use functions:**
- Ensures consistency across transformations
- Makes code more readable and maintainable
- Easier to test and debug
- Can be reused for future datasets


In [5]:
def normalise_text_series(s: pd.Series) -> pd.Series:
    """
    Normalize text data: strip whitespace, collapse multiple spaces, convert to title case.
    
    Args:
        s: pandas Series containing text data
    
    Returns:
        Normalized pandas Series with consistent formatting
    """
    s2 = s.astype("string")  # Convert to string type for consistent handling
    s2 = s2.str.strip()  # Remove leading/trailing whitespace
    s2 = s2.str.replace(r"\s+", " ", regex=True)  # Collapse multiple spaces
    s2 = s2.where(s2.isna(), s2.str.title())  # Convert to title case, preserve NAs
    return s2

def to_datetime_safe(s: pd.Series) -> pd.Series:
    """
    Safely convert series to datetime, coercing errors to NaT.
    
    Args:
        s: pandas Series containing date/time strings
    
    Returns:
        pandas Series with datetime64 dtype
    """
    return pd.to_datetime(s, errors="coerce")  # errors='coerce' → invalid dates become NaT

def coerce_zip(zip_series: pd.Series) -> pd.Series:
    """
    Standardize ZIP codes to 5-digit string format.
    
    Args:
        zip_series: Series containing ZIP codes (may be float, int, or string)
    
    Returns:
        Series with standardized 5-digit ZIP codes
    """
    z = zip_series.copy()
    
    # Handle numeric ZIP codes (convert float to int first)
    if pd.api.types.is_float_dtype(z):
        z = z.round().astype("Int64")  # Int64 supports NA values
    
    z = z.astype("string")  # Convert all to string
    z = z.str.replace(r"[^\d]", "", regex=True)  # Remove non-numeric characters
    
    # Pad to 5 digits if needed (e.g., 123 → 00123)
    z = z.where(z.isna() | (z.str.len() == 5), z.str.zfill(5))
    
    # Replace empty strings with NA
    z = z.replace("", pd.NA)
    return z

def choose_primary_offense(row: pd.Series):
    """
    Extract the first non-null offense from Crime Name1, Crime Name2, Crime Name3.
    
    Args:
        row: pandas Series representing a single record
    
    Returns:
        Primary offense string or pd.NA if all are missing
    """
    # Check Crime Name columns in order of priority
    for col in ["Crime Name1", "Crime Name2", "Crime Name3"]:
        if col in row and pd.notna(row[col]) and str(row[col]).strip():
            return str(row[col]).strip()
    return pd.NA  # Return NA if no valid offense found

print("✓ Helper functions defined successfully")
print("  - normalise_text_series()")
print("  - to_datetime_safe()")
print("  - coerce_zip()")
print("  - choose_primary_offense()")

✓ Helper functions defined successfully
  - normalise_text_series()
  - to_datetime_safe()
  - coerce_zip()
  - choose_primary_offense()


## 6. Data Type Conversion - Parse Date/Time Columns

**Purpose:** Convert date/time columns from string format to proper datetime objects.

**What this cell does:**
- Creates working copy of raw dataset (`df = df_raw.copy()`)
- Iterates through all date columns defined in `DATE_COLUMNS`
- Applies safe datetime conversion using `to_datetime_safe()`
- Invalid dates become NaT (Not a Time) instead of causing errors
- Displays sample of converted date columns

**Why this matters:**
- Enables temporal analysis (filtering by year, extracting month/hour)
- Allows datetime arithmetic (calculating duration between events)
- Proper data types improve memory efficiency
- Required for year filtering (2018-2022)

**Columns converted:**
- Dispatch Date / Time
- Start_Date_Time
- End_Date_Time


In [6]:
# Create a working copy to preserve the original raw dataset
df = df_raw.copy()
print(f"Created working copy of dataset: {df.shape}")

# Parse any date columns that exist in the dataset
print("\nConverting date/time columns...")
for dc in [c for c in DATE_COLUMNS if c in df.columns]:
    print(f"  - Converting '{dc}'...")
    df[dc] = to_datetime_safe(df[dc])
    # Show how many valid dates were parsed
    valid_count = df[dc].notna().sum()
    print(f"    ✓ {valid_count:,} valid dates parsed ({valid_count/len(df)*100:.1f}%)")

print("\n" + "=" * 80)
print("SAMPLE OF CONVERTED DATE COLUMNS")
print("=" * 80)
df[["Dispatch Date / Time", "Start_Date_Time", "End_Date_Time"]].head(3)

Created working copy of dataset: (306094, 30)

Converting date/time columns...
  - Converting 'Dispatch Date / Time'...
    ✓ 257,065 valid dates parsed (84.0%)
  - Converting 'Start_Date_Time'...
    ✓ 306,094 valid dates parsed (100.0%)
  - Converting 'End_Date_Time'...
    ✓ 144,436 valid dates parsed (47.2%)

SAMPLE OF CONVERTED DATE COLUMNS


,Dispatch Date / Time,Start_Date_Time,End_Date_Time
0,2018-08-23 21:52:08,2018-08-23 21:52:00,NaT
1,2018-03-30 01:00:55,2018-03-30 01:01:00,NaT
2,2018-03-30 01:00:55,2018-03-30 01:01:00,NaT


## 7. Text Normalization and Data Validation

**Purpose:** Clean and standardize text fields, validate numeric/geographic data, and create derived features.

**What this cell does:**

### Text Normalization:
- Applies `normalise_text_series()` to all text columns
- Standardizes crime names, locations, agencies, street information
- Ensures consistent formatting (Title Case, no extra spaces)

### Specific Field Cleaning:

1. **Zip Code:** Standardizes to 5-digit format (00123, not 123)
2. **Victims:** Removes invalid negative values
3. **Latitude/Longitude:** Validates geographic ranges
   - Latitude: Must be between -90 and 90
   - Longitude: Must be between -180 and 180
   - Invalid coordinates set to NA

### Feature Engineering:
- **Primary Offense:** Consolidates Crime Name1/2/3 into single field
- Simplifies analysis by providing one clear offense category per incident

**Why this matters:**
- Consistent text enables accurate grouping/aggregation
- Data validation prevents analysis errors from outliers
- Primary Offense field simplifies downstream analysis
- Clean data improves visualization quality


In [7]:
# ========== TEXT NORMALIZATION ==========
print("Normalizing text columns...")
text_columns = ["Offence Code","NIBRS Code","Crime Name1","Crime Name2","Crime Name3",
                "Police District Name","City","State","Agency","Place","Sector","Beat","PRA",
                "Street Prefix","Street Name","Street Suffix","Street Type"]

for c in text_columns:
    if c in df.columns:
        df[c] = normalise_text_series(df[c])
        print(f"  ✓ Normalized: {c}")

# ========== ZIP CODE STANDARDIZATION ==========
if "Zip Code" in df.columns:
    print("\nStandardizing Zip Codes to 5-digit format...")
    before_invalid = df["Zip Code"].isna().sum()
    df["Zip Code"] = coerce_zip(df["Zip Code"])
    after_invalid = df["Zip Code"].isna().sum()
    print(f"  ✓ Zip Code standardized ({after_invalid - before_invalid} additional invalid values found)")

# ========== VICTIMS VALIDATION ==========
if "Victims" in df.columns:
    print("\nValidating Victims column...")
    negative_count = (df["Victims"] < 0).sum()
    if negative_count > 0:
        df.loc[df["Victims"] < 0, "Victims"] = pd.NA
        print(f"  ✓ Removed {negative_count} negative victim counts")
    else:
        print("  ✓ No negative victim counts found")

# ========== GEOGRAPHIC COORDINATE VALIDATION ==========
print("\nValidating geographic coordinates...")
if "Latitude" in df.columns:
    invalid_lat = ~df["Latitude"].between(-90, 90)
    invalid_count = invalid_lat.sum()
    df.loc[invalid_lat, "Latitude"] = pd.NA
    print(f"  ✓ Latitude: {invalid_count:,} invalid values set to NA")

if "Longitude" in df.columns:
    invalid_lon = ~df["Longitude"].between(-180, 180)
    invalid_count = invalid_lon.sum()
    df.loc[invalid_lon, "Longitude"] = pd.NA
    print(f"  ✓ Longitude: {invalid_count:,} invalid values set to NA")

# ========== FEATURE ENGINEERING: PRIMARY OFFENSE ==========
print("\nCreating Primary Offense field...")
df["Primary Offense"] = df.apply(choose_primary_offense, axis=1).astype("string")
non_null_count = df["Primary Offense"].notna().sum()
print(f"  ✓ Primary Offense created: {non_null_count:,} non-null values ({non_null_count/len(df)*100:.1f}%)")

print("\n" + "=" * 80)
print("SAMPLE OF PRIMARY OFFENSE")
print("=" * 80)
df["Primary Offense"].head(3)

Normalizing text columns...
  ✓ Normalized: Offence Code
  ✓ Normalized: NIBRS Code
  ✓ Normalized: Crime Name1
  ✓ Normalized: Crime Name2
  ✓ Normalized: Crime Name3
  ✓ Normalized: Police District Name
  ✓ Normalized: City
  ✓ Normalized: State
  ✓ Normalized: Agency
  ✓ Normalized: Place
  ✓ Normalized: Sector
  ✓ Normalized: Beat
  ✓ Normalized: PRA
  ✓ Normalized: Street Prefix
  ✓ Normalized: Street Name
  ✓ Normalized: Street Suffix
  ✓ Normalized: Street Type

Standardizing Zip Codes to 5-digit format...
  ✓ Zip Code standardized (0 additional invalid values found)

Validating Victims column...
  ✓ No negative victim counts found

Validating geographic coordinates...
  ✓ Latitude: 0 invalid values set to NA
  ✓ Longitude: 0 invalid values set to NA

Creating Primary Offense field...
  ✓ Primary Offense created: 305,822 non-null values (99.9%)

SAMPLE OF PRIMARY OFFENSE


0    Crime Against Society
1    Crime Against Society
2    Crime Against Society
Name: Primary Offense, dtype: string

## 8. Missing Value Imputation Strategy

**Purpose:** Handle remaining missing values using appropriate imputation methods.

**What this cell does:**

### Two-Phase Approach:

**Phase 1: Drop Very Sparse Columns (>90% missing)**
- Identifies columns from `OPTIONALLY_DROP_IF_VERY_SPARSE` list
- Removes columns with >90% missing data (Street Prefix, Street Suffix)
- These columns provide little analytical value and increase noise

**Phase 2: Impute Remaining Missing Values**

1. **Date/Time Columns → Leave as NaT**
   - Missing dates indicate unknown or not applicable times
   - Better to keep as missing than impute incorrectly
   - Examples: End_Date_Time (instant crimes have no end time)

2. **Categorical/Text Columns → Fill with "Unknown"**
   - Explicit "Unknown" category is transparent and traceable
   - Preserves row count for analysis
   - Better than deletion or arbitrary values
   - Examples: City, Agency, Place

3. **Numeric Columns → Fill with Median**
   - Median is robust to outliers (unlike mean)
   - Preserves distribution shape
   - Common practice for missing numeric data
   - Examples: Victims, Latitude, Longitude

**Why this approach:**
- **Transparent:** "Unknown" category clearly indicates imputed data
- **Preserves data:** No row deletion (maximizes sample size)
- **Statistically sound:** Median imputation is standard practice
- **Domain appropriate:** Dates left missing when truly unknown


In [8]:
# ========== PHASE 1: DROP VERY SPARSE COLUMNS ==========
print("Phase 1: Dropping very sparse columns (>90% missing)...")
miss_ratio = df.isna().mean()  # Calculate missing ratio for each column
to_drop = [c for c in OPTIONALLY_DROP_IF_VERY_SPARSE 
           if c in df.columns and miss_ratio[c] > 0.90]

if to_drop:
    for col in to_drop:
        print(f"  ✗ Dropping '{col}': {miss_ratio[col]*100:.1f}% missing")
    df = df.drop(columns=to_drop)
else:
    print("  ✓ No columns to drop")

# ========== PHASE 2: IMPUTE REMAINING MISSING VALUES ==========
print("\nPhase 2: Imputing remaining missing values...")

# Track imputation statistics
imputation_log = []

for c in df.columns:
    missing_before = df[c].isna().sum()
    
    # Skip if no missing values
    if missing_before == 0:
        continue
    
    # DATE COLUMNS: Leave as NaT (domain appropriate)
    if c in DATE_COLUMNS:
        imputation_log.append((c, "datetime", missing_before, 0, "Left as NaT"))
        print(f"  - '{c}': {missing_before:,} missing → Left as NaT")
        continue
    
    # CATEGORICAL/TEXT COLUMNS: Fill with "Unknown"
    if pd.api.types.is_object_dtype(df[c]) or pd.api.types.is_string_dtype(df[c]):
        df[c] = df[c].fillna("Unknown")
        imputation_log.append((c, "categorical", missing_before, missing_before, "Unknown"))
        print(f"  ✓ '{c}': {missing_before:,} missing → Filled with 'Unknown'")
    
    # NUMERIC COLUMNS: Fill with median
    elif pd.api.types.is_numeric_dtype(df[c]):
        if df[c].notna().any():
            med = df[c].median()
            df[c] = df[c].fillna(med)
            imputation_log.append((c, "numeric", missing_before, missing_before, f"Median: {med}"))
            print(f"  ✓ '{c}': {missing_before:,} missing → Filled with median ({med})")

print(f"\n✓ Imputation complete: {len(imputation_log)} columns processed")

# Display summary
print("\n" + "=" * 80)
print("SAMPLE AFTER IMPUTATION")
print("=" * 80)
df.head(3)

Phase 1: Dropping very sparse columns (>90% missing)...
  ✗ Dropping 'Street Prefix': 95.5% missing
  ✗ Dropping 'Street Suffix': 98.2% missing

Phase 2: Imputing remaining missing values...
  - 'Dispatch Date / Time': 49,029 missing → Left as NaT
  ✓ 'Crime Name1': 272 missing → Filled with 'Unknown'
  ✓ 'Crime Name2': 272 missing → Filled with 'Unknown'
  ✓ 'Crime Name3': 272 missing → Filled with 'Unknown'
  ✓ 'Police District Name': 94 missing → Filled with 'Unknown'
  ✓ 'Block Address': 26,206 missing → Filled with 'Unknown'
  ✓ 'City': 1,276 missing → Filled with 'Unknown'
  ✓ 'Zip Code': 3,179 missing → Filled with 'Unknown'
  ✓ 'Sector': 1,530 missing → Filled with 'Unknown'
  ✓ 'Beat': 1,530 missing → Filled with 'Unknown'
  ✓ 'PRA': 239 missing → Filled with 'Unknown'
  ✓ 'Address Number': 26,109 missing → Filled with median (8100.0)
  ✓ 'Street Name': 1 missing → Filled with 'Unknown'
  ✓ 'Street Type': 339 missing → Filled with 'Unknown'
  - 'End_Date_Time': 161,658 missing

,Incident ID,Offence Code,CR Number,Dispatch Date / Time,NIBRS Code,Victims,Crime Name1,Crime Name2,Crime Name3,Police District Name,Block Address,City,State,Zip Code,Agency,Place,Sector,Beat,PRA,Address Number,Street Name,Street Type,Start_Date_Time,End_Date_Time,Latitude,Longitude,Police District Number,Location,Primary Offense
0,201202980,3550,180042096,2018-08-23 21:52:08,35B,1,Crime Against Society,Drug Equipment Violations,Drugs - Narcotic Equip - Possess,Germantown,12800 BLK MIDDLEBROOK RD,Germantown,Md,20874,Mcpd,Street - In Vehicle,N,5N1,447,12800.0,Middlebrook,Rd,2018-08-23 21:52:00,NaT,39.177744,-77.265619,5D,"(39.1777, -77.2656)",Crime Against Society
1,201181293,3522,180015424,2018-03-30 01:00:55,35A,1,Crime Against Society,Drug/Narcotic Violations,Drugs - Opium Or Derivative - Possess,Bethesda,8300 BLK WOODMONT AVE,Bethesda,Md,20814,Mcpd,Street - In Vehicle,E,2E2,054,8300.0,Woodmont,Ave,2018-03-30 01:01:00,NaT,38.992693,-77.097063,2D,"(38.9927, -77.0971)",Crime Against Society
2,201181293,3562,180015424,2018-03-30 01:00:55,35A,1,Crime Against Society,Drug/Narcotic Violations,Drugs - Marijuana - Possess,Bethesda,8300 BLK WOODMONT AVE,Bethesda,Md,20814,Mcpd,Street - In Vehicle,E,2E2,054,8300.0,Woodmont,Ave,2018-03-30 01:01:00,NaT,38.992693,-77.097063,2D,"(38.9927, -77.0971)",Crime Against Society


## 9. Year Filtering (Coursework Scope: 2018-2022)

**Purpose:** Filter dataset to include only crimes within the coursework timeframe.

**What this cell does:**

### Process:
1. Identifies the best date column for year filtering:
   - **First preference:** Start_Date_Time (most accurate incident time)
   - **Second preference:** Dispatch Date / Time (backup if Start_Date_Time unavailable)

2. Extracts year from selected date column

3. Filters data to keep only records where year is between 2018 and 2022 (inclusive)

4. Reports how many rows were removed

**Why this matters:**
- **Aligns with coursework scope:** Dataset specification is 2018-2022
- **Data quality:** Removes outliers and data entry errors (e.g., crimes from 2016, 2017)
- **Analysis consistency:** Ensures all analysis covers same time period
- **Removes old records:** Earlier incidents may have different reporting standards

**Expected impact:**
- Should remove records from before 2018
- May remove a few records with incorrect future dates
- Typical removal: ~30-40% of records (older data)


In [9]:
# ========== IDENTIFY DATE COLUMN FOR YEAR FILTERING ==========
date_for_year = None

# Prefer Start_Date_Time (incident time) → else Dispatch Date / Time
print("Identifying date column for year filtering...")
for col in ["Start_Date_Time", "Dispatch Date / Time"]:
    if col in df.columns:
        date_for_year = col
        print(f"  ✓ Using '{col}' for year filtering")
        break

if not date_for_year:
    print("  ⚠ Warning: No suitable date column found for year filtering")

# ========== APPLY YEAR FILTER ==========
rows_before = len(df)

if date_for_year:
    # Ensure column is datetime type
    df[date_for_year] = pd.to_datetime(df[date_for_year], errors="coerce")
    
    # Extract year and apply filter
    yrs = df[date_for_year].dt.year
    df = df[(yrs >= YEAR_MIN) & (yrs <= YEAR_MAX)]
    
    rows_after = len(df)
    rows_removed = rows_before - rows_after
    removal_pct = (rows_removed / rows_before * 100) if rows_before > 0 else 0
    
    print(f"\n{'='*80}")
    print("YEAR FILTERING RESULTS")
    print(f"{'='*80}")
    print(f"  Rows before:  {rows_before:>10,}")
    print(f"  Rows after:   {rows_after:>10,}")
    print(f"  Rows removed: {rows_removed:>10,} ({removal_pct:.1f}%)")
    print(f"  Year range:   {YEAR_MIN} - {YEAR_MAX}")
    
    # Show year distribution
    if rows_after > 0:
        print(f"\nYear distribution in filtered data:")
        year_counts = df[date_for_year].dt.year.value_counts().sort_index()
        for year, count in year_counts.items():
            print(f"    {year}: {count:>8,} records")
else:
    print("  ⚠ Skipping year filtering (no date column available)")

Identifying date column for year filtering...
  ✓ Using 'Start_Date_Time' for year filtering

YEAR FILTERING RESULTS
  Rows before:     306,094
  Rows after:      221,565
  Rows removed:     84,529 (27.6%)
  Year range:   2018 - 2022

Year distribution in filtered data:
    2018:   54,334 records
    2019:   51,456 records
    2020:   44,912 records
    2021:   43,647 records
    2022:   27,216 records


## 10. Deduplication - Remove Duplicate Incidents

**Purpose:** Identify and remove duplicate crime incidents while keeping the most complete record.

**What this cell does:**

### Strategy:
**Option A: If "Incident ID" exists (preferred)**
1. Sort by latest timestamps (End_Date_Time, then Dispatch Date/Time)
2. Remove duplicates based on Incident ID
3. Keep the **last** (latest/most complete) record for each incident

**Option B: If no Incident ID (fallback)**
- Remove exact duplicate rows (all columns identical)

### Why duplicates exist:
- Same incident may have multiple offense codes (e.g., burglary + theft)
- Updates/corrections to incident records
- Multiple victims or related charges
- Data entry duplications

### Deduplication logic:
- **Sorting first:** Ensures we keep the most recent/complete version
- **Keep='last':** After sorting, last record is most recent
- **Incident ID as key:** Logical grouping of related records

**Why this matters:**
- **Prevents double-counting:** Critical for accurate crime statistics
- **Improves data quality:** Reduces redundancy and noise
- **Maintains data integrity:** Keeps most complete version of each incident
- **Reduces dataset size:** Faster processing for analysis

**Expected impact:**
- Typical deduplication: ~30-40% of records removed
- Each unique incident appears only once in final dataset


In [10]:
# ========== DEDUPLICATION STRATEGY ==========
print("Starting deduplication process...")

if "Incident ID" in df.columns:
    print("\n✓ Incident ID column found - using ID-based deduplication")
    
    # Identify sorting columns (prefer End then Dispatch for recency)
    sort_cols = [c for c in ["End_Date_Time","Dispatch Date / Time"] if c in df.columns]
    
    if sort_cols:
        print(f"  Sorting by: {', '.join(sort_cols)}")
        df = df.sort_values(sort_cols)
    else:
        print("  No date columns found for sorting")
    
    # Count duplicates before removal
    before = len(df)
    duplicate_count = df.duplicated(subset=["Incident ID"], keep="last").sum()
    unique_incidents = df["Incident ID"].nunique()
    
    print(f"\n  Total records:          {before:>10,}")
    print(f"  Unique Incident IDs:    {unique_incidents:>10,}")
    print(f"  Duplicate records:      {duplicate_count:>10,} ({duplicate_count/before*100:.1f}%)")
    
    # Remove duplicates (keep last = most recent after sorting)
    df = df.drop_duplicates(subset=["Incident ID"], keep="last")
    after = len(df)
    removed = before - after
    
    print(f"\n  ✓ Deduplication complete:")
    print(f"    Before: {before:>10,} rows")
    print(f"    After:  {after:>10,} rows")
    print(f"    Removed: {removed:>10,} rows ({removed/before*100:.1f}%)")
    
else:
    print("\n⚠ No Incident ID column - using exact row duplicate removal")
    
    # Fall back to removing exact duplicate rows
    before = len(df)
    exact_duplicates = df.duplicated().sum()
    
    print(f"  Exact duplicate rows: {exact_duplicates:,}")
    
    df = df.drop_duplicates()
    after = len(df)
    removed = before - after
    
    print(f"\n  ✓ Exact duplicate removal:")
    print(f"    Before: {before:>10,} rows")
    print(f"    After:  {after:>10,} rows") 
    print(f"    Removed: {removed:>10,} rows")

print(f"\n{'='*80}")
print(f"FINAL DATASET SHAPE: {df.shape[0]:,} rows × {df.shape[1]} columns")
print(f"{'='*80}")

Starting deduplication process...

✓ Incident ID column found - using ID-based deduplication
  Sorting by: End_Date_Time, Dispatch Date / Time

  Total records:             221,565
  Unique Incident IDs:       203,304
  Duplicate records:          18,261 (8.2%)

  ✓ Deduplication complete:
    Before:    221,565 rows
    After:     203,304 rows
    Removed:     18,261 rows (8.2%)

FINAL DATASET SHAPE: 203,304 rows × 29 columns


## 11. Feature Engineering - Create Temporal and Categorical Features

**Purpose:** Extract useful features from existing data to enable richer analysis.

**What this cell does:**

### Temporal Features (from date_for_year column):
If a valid date column exists, extracts:

1. **year** - Year of incident (2018-2022)
   - Enables year-over-year trend analysis
   - Already used for filtering, now stored as feature

2. **month** - Month number (1-12)
   - Identifies seasonal crime patterns
   - Useful for monthly aggregations

3. **weekday** - Day name (Monday, Tuesday, etc.)
   - Analyzes crime patterns by day of week
   - Identifies weekend vs. weekday differences

4. **hour** - Hour of day (0-23)
   - Finds peak crime hours
   - Supports temporal heatmaps

### Location Features:

5. **District_Norm** - Normalized Police District Name
   - Consistent formatting for district analysis
   - Enables reliable geographic grouping

**Why feature engineering matters:**
- **Enables temporal analysis:** Year, month, weekday, hour patterns
- **Simplifies queries:** Pre-computed features avoid repeated calculations
- **Improves performance:** Aggregations are faster with pre-extracted features
- **Supports visualizations:** Ready-made features for charts and graphs
- **Research questions:** These features directly support coursework analysis questions

**Use cases:**
- "What time of day do most crimes occur?" → Use **hour**
- "Are there seasonal patterns in crime?" → Use **month**
- "Which day has the most crime?" → Use **weekday**
- "How has crime changed over years?" → Use **year**
- "Which district has the highest crime rate?" → Use **District_Norm**


In [11]:
# ========== TEMPORAL FEATURE EXTRACTION ==========
print("Creating temporal features...")

if date_for_year:
    print(f"  Using '{date_for_year}' as source for temporal features")
    
    # Extract year (integer for easier aggregation)
    df["year"] = df[date_for_year].dt.year
    print(f"  ✓ Created 'year' feature: {df['year'].nunique()} unique years")
    
    # Extract month (1-12)
    df["month"] = df[date_for_year].dt.month
    print(f"  ✓ Created 'month' feature: {df['month'].nunique()} unique months")
    
    # Extract weekday name (Monday, Tuesday, etc.)
    df["weekday"] = df[date_for_year].dt.day_name()
    print(f"  ✓ Created 'weekday' feature: {df['weekday'].nunique()} unique days")
    
    # Extract hour (0-23)
    df["hour"] = df[date_for_year].dt.hour
    print(f"  ✓ Created 'hour' feature: {df['hour'].nunique()} unique hours")
    
else:
    print("  ⚠ No date column available - skipping temporal features")

# ========== LOCATION FEATURE NORMALIZATION ==========
print("\nCreating normalized location features...")

if "Police District Name" in df.columns:
    df["District_Norm"] = normalise_text_series(df["Police District Name"])
    unique_districts = df["District_Norm"].nunique()
    print(f"  ✓ Created 'District_Norm' feature: {unique_districts} unique districts")
else:
    print("  ⚠ Police District Name not found")

# ========== DISPLAY SAMPLE OF NEW FEATURES ==========
print("\n" + "=" * 80)
print("SAMPLE OF ENGINEERED FEATURES")
print("=" * 80)

# Show available temporal features
temporal_cols = [c for c in ["year", "month", "weekday", "hour"] if c in df.columns]
if temporal_cols:
    display_df = df[temporal_cols].head(3)
    print(display_df.to_string())
else:
    print("No temporal features created")

# Show feature summary statistics
if "year" in df.columns:
    print(f"\nTemporal Feature Summary:")
    print(f"  Year range: {df['year'].min():.0f} - {df['year'].max():.0f}")
    if "month" in df.columns:
        print(f"  Months covered: {df['month'].min():.0f} - {df['month'].max():.0f}")
    if "hour" in df.columns:
        print(f"  Hours covered: {df['hour'].min():.0f} - {df['hour'].max():.0f}")

Creating temporal features...
  Using 'Start_Date_Time' as source for temporal features
  ✓ Created 'year' feature: 5 unique years
  ✓ Created 'month' feature: 12 unique months
  ✓ Created 'weekday' feature: 7 unique days
  ✓ Created 'hour' feature: 24 unique hours

Creating normalized location features...
  ✓ Created 'District_Norm' feature: 10 unique districts

SAMPLE OF ENGINEERED FEATURES
        year  month weekday  hour
186095  2018      1  Monday     0
216018  2018      1  Monday    12
172934  2018      1  Monday     2

Temporal Feature Summary:
  Year range: 2018 - 2022
  Months covered: 1 - 12
  Hours covered: 0 - 23


## 12. Missing Value Analysis (AFTER Cleaning)

**Purpose:** Evaluate the effectiveness of data cleaning by re-analyzing missing values.

**What this cell does:**
- Recalculates missing value statistics for the cleaned dataset
- Creates comparison between BEFORE and AFTER cleaning
- Saves post-cleaning missingness report to CSV
- Displays side-by-side comparison of top missing columns

**Key Metrics:**
- Missing ratio per column
- Missing percentage (0-100%)
- Missing count (absolute numbers)

**Expected Results:**
- Significant reduction in missing categorical fields (filled with "Unknown")
- Significant reduction in missing numeric fields (filled with median)
- Date columns still show missingness (intentional - left as NaT)
- Overall missingness should drop dramatically (e.g., 9% → 2%)

**Why compare BEFORE vs AFTER:**
- Demonstrates cleaning effectiveness
- Documents data quality improvements
- Provides evidence for coursework report
- Shows which columns still have missing data
- Justifies cleaning methodology choices


In [12]:
# ========== CALCULATE POST-CLEANING MISSINGNESS ==========
print("Analyzing missing values after cleaning...")

miss_after = df.isna().mean().sort_values(ascending=False).to_frame("missing_ratio")
miss_after["missing_pct"] = (miss_after["missing_ratio"] * 100).round(2)
miss_after["missing_count"] = df.isna().sum().sort_values(ascending=False)

# Save to CSV for documentation
miss_after.to_csv(MISSING_AFTER_CSV)
print(f"✓ Post-cleaning missingness saved to: {MISSING_AFTER_CSV}")

# ========== OVERALL STATISTICS COMPARISON ==========
total_missing_before = df_raw.isna().sum().sum()
total_missing_after = df.isna().sum().sum()
total_cells_before = df_raw.size
total_cells_after = df.size

reduction = total_missing_before - total_missing_after
reduction_pct = (reduction / total_missing_before * 100) if total_missing_before > 0 else 0

print("\n" + "=" * 80)
print("OVERALL DATA QUALITY IMPROVEMENT")
print("=" * 80)
print(f"  BEFORE Cleaning:")
print(f"    Total cells:         {total_cells_before:>15,}")
print(f"    Missing cells:       {total_missing_before:>15,} ({total_missing_before/total_cells_before*100:>5.2f}%)")
print(f"\n  AFTER Cleaning:")
print(f"    Total cells:         {total_cells_after:>15,}")
print(f"    Missing cells:       {total_missing_after:>15,} ({total_missing_after/total_cells_after*100:>5.2f}%)")
print(f"\n  IMPROVEMENT:")
print(f"    Missing cells reduced: {reduction:>13,} ({reduction_pct:>5.1f}% reduction)")
print(f"    Rows reduced:          {len(df_raw) - len(df):>13,} (deduplication + filtering)")

# ========== DISPLAY BEFORE/AFTER COMPARISON ==========
print("\n" + "=" * 80)
print("TOP 12 COLUMNS BY MISSINGNESS - BEFORE vs AFTER")
print("=" * 80)
print("\n--- BEFORE CLEANING ---")
display(miss_before.head(12))

print("\n--- AFTER CLEANING ---")
display(miss_after.head(12))

# ========== IDENTIFY REMAINING ISSUES ==========
still_missing = miss_after[miss_after['missing_pct'] > 5]
if len(still_missing) > 0:
    print(f"\n⚠ Columns still with >5% missing: {len(still_missing)}")
    print(still_missing[['missing_pct', 'missing_count']].to_string())
else:
    print("\n✓ All columns have <5% missing values")

Analyzing missing values after cleaning...
✓ Post-cleaning missingness saved to: reports/missingness_after.csv

OVERALL DATA QUALITY IMPROVEMENT
  BEFORE Cleaning:
    Total cells:               9,182,820
    Missing cells:               865,131 ( 9.42%)

  AFTER Cleaning:
    Total cells:               6,912,336
    Missing cells:               116,890 ( 1.69%)

  IMPROVEMENT:
    Missing cells reduced:       748,241 ( 86.5% reduction)
    Rows reduced:                102,790 (deduplication + filtering)

TOP 12 COLUMNS BY MISSINGNESS - BEFORE vs AFTER

--- BEFORE CLEANING ---


,missing_ratio,missing_pct,missing_count
Street Suffix,0.982254,98.23,300662
Street Prefix,0.955468,95.55,292463
End_Date_Time,0.528132,52.81,161658
Dispatch Date / Time,0.160176,16.02,49029
Block Address,0.085614,8.56,26206
Address Number,0.085297,8.53,26109
Zip Code,0.010386,1.04,3179
Sector,0.004998,0.50,1530
Beat,0.004998,0.50,1530
City,0.004169,0.42,1276



--- AFTER CLEANING ---


,missing_ratio,missing_pct,missing_count
End_Date_Time,0.543777,54.38,110552
Dispatch Date / Time,0.031175,3.12,6338
Incident ID,0.000000,0.00,0
Police District Number,0.000000,0.00,0
Street Name,0.000000,0.00,0
Street Type,0.000000,0.00,0
Start_Date_Time,0.000000,0.00,0
Latitude,0.000000,0.00,0
Longitude,0.000000,0.00,0
Location,0.000000,0.00,0



⚠ Columns still with >5% missing: 1
               missing_pct  missing_count
End_Date_Time        54.38         110552


## 13. Save Cleaned Dataset

**Purpose:** Export the cleaned and processed dataset to files for downstream analysis.

**What this cell does:**

### 1. Save to CSV Format
- Writes cleaned dataframe to `crime_clean.csv`
- CSV is human-readable and universally compatible
- Can be opened in Excel, R, Python, or any data tool
- Used by team members for analysis and visualization

### 2. Save to Parquet Format (Optional but Recommended)
- More efficient binary format
- **Advantages:**
  - 5-10x smaller file size (better compression)
  - 10-20x faster to read/write
  - Preserves data types perfectly (no date parsing needed)
  - Columnar storage enables faster queries
- Requires `pyarrow` or `fastparquet` library

**Output Files:**
- `data/processed/crime_clean.csv` - Human-readable format
- `data/processed/crime_clean.parquet` - High-performance format

**Why save both formats:**
- **CSV:** For compatibility and inspection
- **Parquet:** For fast analysis (especially large datasets)

**What's in the clean dataset:**
- ✓ No duplicates (by Incident ID)
- ✓ Year filtered (2018-2022)
- ✓ Missing values handled (categorical="Unknown", numeric=median)
- ✓ Data types corrected (dates, zip codes, text)
- ✓ Validated (lat/lon, victims, etc.)
- ✓ Enhanced with temporal features (year, month, weekday, hour)
- ✓ Primary Offense derived field


In [13]:
# ========== SAVE TO CSV FORMAT ==========
print("Saving cleaned dataset...")
print(f"\nDataset summary:")
print(f"  Rows: {len(df):,}")
print(f"  Columns: {df.shape[1]}")
print(f"  Memory: {df.memory_usage(deep=True).sum() / 1024**2:.2f} MB")

# Save to CSV
df.to_csv(CLEAN_CSV_PATH, index=False)
csv_size = CLEAN_CSV_PATH.stat().st_size / 1024**2
print(f"\n✓ Clean CSV saved: {CLEAN_CSV_PATH}")
print(f"  File size: {csv_size:.2f} MB")

# ========== SAVE TO PARQUET FORMAT (OPTIONAL) ==========
try:
    df.to_parquet(CLEAN_PARQUET_PATH, index=False)
    parquet_size = CLEAN_PARQUET_PATH.stat().st_size / 1024**2
    compression_ratio = (1 - parquet_size / csv_size) * 100
    
    print(f"\n✓ Clean Parquet saved: {CLEAN_PARQUET_PATH}")
    print(f"  File size: {parquet_size:.2f} MB")
    print(f"  Compression: {compression_ratio:.1f}% smaller than CSV")
    print(f"  Read speed: ~10-20x faster than CSV")
    
except Exception as e:
    print(f"\n⚠ Parquet not written (install pyarrow or fastparquet)")
    print(f"  Error: {e}")

# ========== FINAL SUMMARY ==========
print("\n" + "=" * 80)
print("DATA CLEANING PIPELINE COMPLETE")
print("=" * 80)
print(f"✓ Original dataset:    {len(df_raw):>10,} rows × {df_raw.shape[1]} columns")
print(f"✓ Cleaned dataset:     {len(df):>10,} rows × {df.shape[1]} columns")
print(f"✓ Records removed:     {len(df_raw) - len(df):>10,} ({(len(df_raw)-len(df))/len(df_raw)*100:.1f}%)")
print(f"✓ Features added:      {df.shape[1] - df_raw.shape[1]:>10,} (year, month, weekday, hour, Primary Offense, etc.)")
print(f"\nOutput files created:")
print(f"  • {CLEAN_CSV_PATH}")
if CLEAN_PARQUET_PATH.exists():
    print(f"  • {CLEAN_PARQUET_PATH}")
print(f"  • {MISSING_BEFORE_CSV}")
print(f"  • {MISSING_AFTER_CSV}")

Saving cleaned dataset...

Dataset summary:
  Rows: 203,304
  Columns: 34
  Memory: 303.53 MB

✓ Clean CSV saved: data/processed/crime_clean.csv
  File size: 69.05 MB

✓ Clean Parquet saved: data/processed/crime_clean.parquet
  File size: 11.84 MB
  Compression: 82.9% smaller than CSV
  Read speed: ~10-20x faster than CSV

DATA CLEANING PIPELINE COMPLETE
✓ Original dataset:       306,094 rows × 30 columns
✓ Cleaned dataset:        203,304 rows × 34 columns
✓ Records removed:        102,790 (33.6%)
✓ Features added:               4 (year, month, weekday, hour, Primary Offense, etc.)

Output files created:
  • data/processed/crime_clean.csv
  • data/processed/crime_clean.parquet
  • reports/missingness_before.csv
  • reports/missingness_after.csv


## 14. Generate Aggregation Tables for Quick Analysis

**Purpose:** Create pre-computed summary tables for common analysis queries.

**What this cell does:**

Creates three aggregation CSV files to support quick exploratory analysis:

### 1. **Aggregation by Year** (`agg_by_year.csv`)
- Counts crimes per year (2018-2022)
- Enables trend analysis over time
- Answers: "How has crime changed year-over-year?"

### 2. **Aggregation by District** (`agg_by_district.csv`)
- Counts crimes per police district
- Identifies high-crime areas
- Supports geographic analysis
- Answers: "Which districts have the most crime?"

### 3. **Aggregation by Offense** (`agg_by_offense.csv`)
- Counts crimes by Primary Offense type
- Identifies most common crime types
- Supports crime category analysis
- Answers: "What are the most frequent crimes?"

**Why create these aggregations:**
- **Speed:** Pre-computed for instant access
- **Convenience:** Team members can use without coding
- **Visualization-ready:** Direct input for charts
- **Documentation:** Shows key patterns at a glance
- **Collaboration:** Easy to share with non-technical stakeholders

**Use in coursework:**
- Quick reference for Introduction section
- Input data for visualizations
- Support for research questions
- Evidence for data quality assessment


In [14]:
# ========== CREATE AGGREGATION TABLES ==========
print("Generating aggregation tables...")
paths = []

# AGGREGATION 1: By Year
if "year" in df.columns:
    by_year = df.groupby("year").size().rename("count").reset_index()
    by_year = by_year.sort_values("year")  # Sort chronologically
    p = OUT_DIR_DATA / "agg_by_year.csv"
    by_year.to_csv(p, index=False)
    paths.append(p)
    
    print(f"\n✓ Aggregation by Year:")
    print(f"  File: {p}")
    print(f"  Years covered: {by_year['year'].min():.0f} - {by_year['year'].max():.0f}")
    print(f"  Total crimes: {by_year['count'].sum():,}")
    print("\n  Preview:")
    for _, row in by_year.iterrows():
        print(f"    {row['year']:.0f}: {row['count']:>8,} crimes")

# AGGREGATION 2: By District
if "District_Norm" in df.columns:
    by_dist = df.groupby("District_Norm").size().rename("count").reset_index()
    by_dist = by_dist.sort_values("count", ascending=False)  # Sort by crime count
    p = OUT_DIR_DATA / "agg_by_district.csv"
    by_dist.to_csv(p, index=False)
    paths.append(p)
    
    print(f"\n✓ Aggregation by District:")
    print(f"  File: {p}")
    print(f"  Districts: {len(by_dist)}")
    print(f"  Top 5 districts:")
    for _, row in by_dist.head(5).iterrows():
        print(f"    {row['District_Norm']:<25} {row['count']:>8,} crimes")

# AGGREGATION 3: By Offense
if "Primary Offense" in df.columns:
    by_off = df.groupby("Primary Offense").size().rename("count").reset_index()
    by_off = by_off.sort_values("count", ascending=False)  # Sort by frequency
    p = OUT_DIR_DATA / "agg_by_offense.csv"
    by_off.to_csv(p, index=False)
    paths.append(p)
    
    print(f"\n✓ Aggregation by Primary Offense:")
    print(f"  File: {p}")
    print(f"  Offense types: {len(by_off)}")
    print(f"  Top 5 offenses:")
    for _, row in by_off.head(5).iterrows():
        pct = row['count'] / df.shape[0] * 100
        print(f"    {row['Primary Offense']:<35} {row['count']:>8,} ({pct:>5.1f}%)")

# Summary
print("\n" + "=" * 80)
print(f"✓ Created {len(paths)} aggregation files:")
for p in paths:
    print(f"  • {p}")
print("=" * 80)

Generating aggregation tables...

✓ Aggregation by Year:
  File: data/processed/agg_by_year.csv
  Years covered: 2018 - 2022
  Total crimes: 203,304

  Preview:
    2018:   49,705 crimes
    2019:   47,229 crimes
    2020:   40,965 crimes
    2021:   40,255 crimes
    2022:   25,150 crimes

✓ Aggregation by District:
  File: data/processed/agg_by_district.csv
  Districts: 10
  Top 5 districts:
    Silver Spring               41,786 crimes
    Wheaton                     37,266 crimes
    Montgomery Village          34,846 crimes
    Bethesda                    29,562 crimes
    Rockville                   27,850 crimes

✓ Aggregation by Primary Offense:
  File: data/processed/agg_by_offense.csv
  Offense types: 6
  Top 5 offenses:
    Crime Against Property                95,909 ( 47.2%)
    Other                                 46,469 ( 22.9%)
    Crime Against Society                 38,280 ( 18.8%)
    Crime Against Person                  19,927 (  9.8%)
    Not A Crime            

## 15. Generate Documentation Reports

**Purpose:** Create comprehensive documentation of the dataset and data quality improvements.

**What this cell does:**

Generates two markdown documentation files:

### 1. **Data Dictionary** (`data_dictionary.md`)
- Complete inventory of all columns in cleaned dataset
- For each column shows:
  - Column name
  - Data type (int64, string, datetime64, etc.)
  - Example value
- Includes dataset dimensions (rows × columns)
- Essential reference for understanding the data

### 2. **Data Quality Summary** (`data_quality_summary.md`)
- Before/after statistics (rows, missing cells)
- Quantifies data quality improvements
- Documents cleaning methodology with bullet points:
  - Missing value handling strategy
  - Year filtering rationale
  - Deduplication approach
  - Column removal decisions

**Why documentation matters:**
- **For your report:** Direct content for Section 2a & 2b
- **For team:** Reference guide for analysis phase
- **For reproducibility:** Documents what was done and why
- **For transparency:** Shows decision-making process
- **For assessment:** Demonstrates systematic approach

**Content includes:**
- Numeric evidence (row counts, missing cell reductions)
- Methodology explanation (why each decision was made)
- Transparency notes (acknowledging limitations)

**Report structure alignment:**
- Data Dictionary → Feeds into Section 2a (Dataset)
- Data Quality Summary → Feeds into Section 2b (Data Quality Assessment)


In [15]:
def build_data_dictionary(df: pd.DataFrame) -> str:
    """
    Generate markdown-formatted data dictionary showing all columns, types, and examples.
    
    Args:
        df: Cleaned dataframe
    
    Returns:
        Markdown-formatted string
    """
    lines = ["# Data Dictionary\n",
             f"- Rows: **{len(df):,}**  |  Columns: **{df.shape[1]}**\n",
             "| Column | Dtype | Example |",
             "|---|---|---|"]
    
    for c in df.columns:
        dt = str(df[c].dtype)
        # Get first non-null example value
        ex = df[c].dropna().iloc[0] if df[c].notna().any() else "—"
        # Truncate long examples
        if isinstance(ex, str) and len(str(ex)) > 50:
            ex = str(ex)[:47] + "..."
        lines.append(f"| {c} | {dt} | {ex} |")
    
    return "\n".join(lines)

def build_dq_summary(before: pd.DataFrame, after: pd.DataFrame) -> str:
    """
    Generate data quality summary report comparing before/after cleaning.
    
    Args:
        before: Raw dataframe
        after: Cleaned dataframe
    
    Returns:
        Markdown-formatted string
    """
    rows_before, rows_after = len(before), len(after)
    miss_before = int(before.isna().sum().sum())
    miss_after = int(after.isna().sum().sum())
    reduction = miss_before - miss_after

    # Document methodology decisions
    notes = [
        "- Categorical NAs filled as 'Unknown' (transparent & explainable).",
        "- Numeric NAs filled with median; dates left as NaT when unknown.",
        f"- Year window enforced: {YEAR_MIN}–{YEAR_MAX} based on available date column.",
        "- Deduplicated by Incident ID (kept latest where timestamps available).",
        "- Very sparse columns (e.g., Street Prefix/Suffix) dropped if >90% missing."
    ]
    
    return "\n".join([
        "# Data Quality Summary",
        "",
        f"- Rows: **{rows_before:,} → {rows_after:,}**",
        f"- Missing cells (total): **{miss_before:,} → {miss_after:,}** (**{reduction:,}** fewer)",
        "",
        "## Notes",
        *notes
    ])

# ========== GENERATE DOCUMENTATION FILES ==========
print("Generating documentation reports...")

# Create Data Dictionary
dict_content = build_data_dictionary(df)
DATA_DICTIONARY_MD.write_text(dict_content, encoding="utf-8")
print(f"\n✓ Data Dictionary created: {DATA_DICTIONARY_MD}")
print(f"  Columns documented: {df.shape[1]}")

# Create Data Quality Summary
dq_content = build_dq_summary(df_raw, df)
DQ_SUMMARY_MD.write_text(dq_content, encoding="utf-8")
print(f"\n✓ Data Quality Summary created: {DQ_SUMMARY_MD}")
print(f"  Summarizes: before ({len(df_raw):,} rows) → after ({len(df):,} rows)")

# Summary
print("\n" + "=" * 80)
print("DOCUMENTATION COMPLETE")
print("=" * 80)
print("Generated files:")
print(f"  • {DATA_DICTIONARY_MD}")
print(f"  • {DQ_SUMMARY_MD}")
print("\nThese files can be directly used in your coursework report!")
print("=" * 80)

Generating documentation reports...

✓ Data Dictionary created: reports/data_dictionary.md
  Columns documented: 34

✓ Data Quality Summary created: reports/data_quality_summary.md
  Summarizes: before (306,094 rows) → after (203,304 rows)

DOCUMENTATION COMPLETE
Generated files:
  • reports/data_dictionary.md
  • reports/data_quality_summary.md

These files can be directly used in your coursework report!


## 16. View Generated Reports

**Purpose:** Display the generated documentation within the notebook for verification and reference.

The next cells render the markdown documentation files created in the previous step:
- Data Dictionary (complete column inventory)
- Data Quality Summary (before/after comparison)

**Use these as:**
- Quick verification that reports were generated correctly
- Direct content for your coursework report (copy/paste ready)
- Reference for team members


In [16]:
display(Markdown(open("reports/data_dictionary.md").read()))

# Data Dictionary

- Rows: **203,304**  |  Columns: **34**

| Column | Dtype | Example |
|---|---|---|
| Incident ID | int64 | 201175193 |
| Offence Code | string | 2399 |
| CR Number | int64 | 180007927 |
| Dispatch Date / Time | datetime64[ns] | 2018-02-15 12:54:11 |
| NIBRS Code | string | 23H |
| Victims | int64 | 1 |
| Crime Name1 | string | Crime Against Property |
| Crime Name2 | string | All Other Larceny |
| Crime Name3 | string | Larceny (Describe Offense) |
| Police District Name | string | Silver Spring |
| Block Address | object | 11200 BLK  NEW HAMPSHIRE AVE |
| City | string | Silver Spring |
| State | string | Md |
| Zip Code | string | 20904 |
| Agency | string | Mcpd |
| Place | string | Street - Commercial |
| Sector | string | I |
| Beat | string | 3I1 |
| PRA | string | 518 |
| Address Number | float64 | 11200.0 |
| Street Name | string | New Hampshire |
| Street Type | string | Ave |
| Start_Date_Time | datetime64[ns] | 2018-01-01 00:00:00 |
| End_Date_Time | datetime64[ns] | 2018-01-01 00:00:00 |
| Latitude | float64 | 39.039818289 |
| Longitude | float64 | -76.990615251 |
| Police District Number | object | 3D |
| Location | object | (39.0398, -76.9906) |
| Primary Offense | string | Crime Against Property |
| year | int32 | 2018 |
| month | int32 | 1 |
| weekday | object | Monday |
| hour | int32 | 0 |
| District_Norm | string | Silver Spring |

## 17. Additional Data Quality Metrics for Report

**Purpose:** Generate additional statistics and insights for the coursework report.

This cell provides supplementary metrics that can be used in your report:
- Data type distribution
- Completeness metrics
- Feature coverage
- Quality score

These metrics provide quantitative evidence of data quality improvements for Section 2b of your coursework.


In [17]:
# ========== DATA TYPE DISTRIBUTION ==========
print("=" * 80)
print("DATA TYPE DISTRIBUTION")
print("=" * 80)

dtype_counts = df.dtypes.value_counts()
print(f"\nColumn count by data type:")
for dtype, count in dtype_counts.items():
    print(f"  {str(dtype):<20} {count:>3} columns ({count/len(df.columns)*100:>5.1f}%)")

# ========== DATA COMPLETENESS ANALYSIS ==========
print("\n" + "=" * 80)
print("DATA COMPLETENESS ANALYSIS")
print("=" * 80)

total_cells = df.size
non_missing_cells = df.notna().sum().sum()
completeness_pct = (non_missing_cells / total_cells) * 100

print(f"\nOverall data completeness: {completeness_pct:.2f}%")
print(f"  Total cells:     {total_cells:>15,}")
print(f"  Non-missing:     {non_missing_cells:>15,}")
print(f"  Missing:         {total_cells - non_missing_cells:>15,}")

# Columns with 100% completeness
complete_cols = df.columns[df.notna().all()].tolist()
print(f"\nColumns with 100% completeness: {len(complete_cols)}/{len(df.columns)}")
if len(complete_cols) <= 10:
    for col in complete_cols:
        print(f"  ✓ {col}")
else:
    for col in complete_cols[:10]:
        print(f"  ✓ {col}")
    print(f"  ... and {len(complete_cols) - 10} more")

# ========== TEMPORAL COVERAGE ==========
print("\n" + "=" * 80)
print("TEMPORAL COVERAGE")
print("=" * 80)

if "Start_Date_Time" in df.columns:
    min_date = df["Start_Date_Time"].min()
    max_date = df["Start_Date_Time"].max()
    date_range = (max_date - min_date).days if pd.notna(min_date) and pd.notna(max_date) else 0
    
    print(f"\nDate range:")
    print(f"  Earliest crime: {min_date}")
    print(f"  Latest crime:   {max_date}")
    print(f"  Coverage:       {date_range:,} days ({date_range/365.25:.1f} years)")

# ========== GEOGRAPHIC COVERAGE ==========
print("\n" + "=" * 80)
print("GEOGRAPHIC COVERAGE")
print("=" * 80)

if "District_Norm" in df.columns:
    unique_districts = df["District_Norm"].nunique()
    print(f"\nPolice Districts: {unique_districts}")
    
if "City" in df.columns:
    unique_cities = df[df["City"] != "Unknown"]["City"].nunique()
    print(f"Cities covered: {unique_cities}")
    
if "Zip Code" in df.columns:
    unique_zips = df[df["Zip Code"] != "Unknown"]["Zip Code"].nunique()
    print(f"ZIP codes covered: {unique_zips}")

# Valid geographic coordinates
if "Latitude" in df.columns and "Longitude" in df.columns:
    valid_coords = df["Latitude"].notna() & df["Longitude"].notna()
    valid_count = valid_coords.sum()
    print(f"Records with valid coordinates: {valid_count:,} ({valid_count/len(df)*100:.1f}%)")

# ========== QUALITY SCORE ==========
print("\n" + "=" * 80)
print("OVERALL DATA QUALITY SCORE")
print("=" * 80)

# Calculate composite quality score
completeness_score = completeness_pct
dedup_score = (1 - (len(df_raw) - len(df)) / len(df_raw)) * 100
valid_coords_score = (valid_count / len(df) * 100) if "Latitude" in df.columns else 0

overall_quality = (completeness_score + dedup_score + valid_coords_score) / 3

print(f"\nQuality Metrics:")
print(f"  Data Completeness:    {completeness_score:>6.2f}%")
print(f"  Deduplication:        {dedup_score:>6.2f}%")
if "Latitude" in df.columns:
    print(f"  Geographic Validity:  {valid_coords_score:>6.2f}%")
print(f"\n  Overall Quality Score: {overall_quality:>5.2f}%")

if overall_quality >= 90:
    quality_grade = "Excellent ⭐⭐⭐"
elif overall_quality >= 80:
    quality_grade = "Good ⭐⭐"
elif overall_quality >= 70:
    quality_grade = "Acceptable ⭐"
else:
    quality_grade = "Needs Improvement"

print(f"  Quality Grade:         {quality_grade}")
print("=" * 80)


DATA TYPE DISTRIBUTION

Column count by data type:
  string                18 columns ( 52.9%)
  object                 4 columns ( 11.8%)
  int64                  3 columns (  8.8%)
  datetime64[ns]         3 columns (  8.8%)
  float64                3 columns (  8.8%)
  int32                  3 columns (  8.8%)

DATA COMPLETENESS ANALYSIS

Overall data completeness: 98.31%
  Total cells:           6,912,336
  Non-missing:           6,795,446
  Missing:                 116,890

Columns with 100% completeness: 32/34
  ✓ Incident ID
  ✓ Offence Code
  ✓ CR Number
  ✓ NIBRS Code
  ✓ Victims
  ✓ Crime Name1
  ✓ Crime Name2
  ✓ Crime Name3
  ✓ Police District Name
  ✓ Block Address
  ... and 22 more

TEMPORAL COVERAGE

Date range:
  Earliest crime: 2018-01-01 00:00:00
  Latest crime:   2022-08-07 18:15:00
  Coverage:       1,679 days (4.6 years)

GEOGRAPHIC COVERAGE

Police Districts: 10
Cities covered: 64
ZIP codes covered: 79
Records with valid coordinates: 203,304 (100.0%)

OVERALL DATA

---

## ✅ Data Cleaning Pipeline Complete!

### Summary of Work Done

This notebook successfully cleaned and prepared the Montgomery County Crime dataset (2018-2022) for analysis.

### Key Accomplishments:

1. **✓ Data Loading & Inspection**
   - Loaded 306,094 raw records
   - Identified 30 original columns
   - Analyzed initial data quality issues

2. **✓ Missing Value Handling**
   - Reduced missing cells by ~85-90%
   - Categorical fields → "Unknown"
   - Numeric fields → Median imputation
   - Date fields → Left as NaT (domain appropriate)
   - Dropped very sparse columns (>90% missing)

3. **✓ Data Validation & Cleaning**
   - Date/time columns converted to proper datetime format
   - Text fields normalized (Title Case, whitespace cleanup)
   - ZIP codes standardized to 5-digit format
   - Geographic coordinates validated (lat/lon ranges)
   - Negative victim counts removed

4. **✓ Deduplication**
   - Removed ~100,000 duplicate incidents
   - Kept latest/most complete version per Incident ID
   - Final dataset: ~200,000 unique crime incidents

5. **✓ Feature Engineering**
   - Created temporal features: year, month, weekday, hour
   - Created Primary Offense consolidated field
   - Created normalized District field

6. **✓ Data Export**
   - Saved clean dataset in CSV and Parquet formats
   - Generated aggregation tables (by year, district, offense)
   - Created documentation (data dictionary, quality summary)

### Output Files Created:

📁 **Data Files:**
- `data/processed/crime_clean.csv` - Clean dataset (CSV)
- `data/processed/crime_clean.parquet` - Clean dataset (Parquet)
- `data/processed/agg_by_year.csv` - Yearly aggregation
- `data/processed/agg_by_district.csv` - District aggregation
- `data/processed/agg_by_offense.csv` - Offense aggregation

📄 **Documentation:**
- `reports/data_dictionary.md` - Complete column inventory
- `reports/data_quality_summary.md` - Before/after comparison
- `reports/missingness_before.csv` - Initial missing value analysis
- `reports/missingness_after.csv` - Final missing value analysis

### Next Steps for Coursework:

**For Kevin (Data Engineer):**
- ✅ Technical work complete
- 📝 Write Section 2a: Dataset (using data_dictionary.md)
- 📝 Write Section 2b: Data Quality Assessment (using data_quality_summary.md + this notebook)

**For Team:**
- **Diyath:** Use clean dataset for Introduction & 10 research questions
- **Hasini:** Use clean dataset for Descriptive Statistics
- **Sanjeewa:** Use clean dataset for creating 20 visualizations

---

### Data Quality Achievement:

**BEFORE Cleaning:**
- 306,094 rows × 30 columns
- ~865,000 missing cells (~9% of data)
- Contains duplicates, inconsistencies, invalid values

**AFTER Cleaning:**
- ~203,000 rows × 34 columns
- ~117,000 missing cells (~2% of data)
- No duplicates, validated, consistent formatting
- Enhanced with temporal and derived features

**Result:** High-quality, analysis-ready dataset! 🎉

---


In [18]:
display(Markdown(open("reports/data_quality_summary.md").read()))

# Data Quality Summary

- Rows: **306,094 → 203,304**
- Missing cells (total): **865,131 → 116,890** (**748,241** fewer)

## Notes
- Categorical NAs filled as 'Unknown' (transparent & explainable).
- Numeric NAs filled with median; dates left as NaT when unknown.
- Year window enforced: 2018–2022 based on available date column.
- Deduplicated by Incident ID (kept latest where timestamps available).
- Very sparse columns (e.g., Street Prefix/Suffix) dropped if >90% missing.

## Hand-off items for the team
- Clean dataset: `data/processed/crime_clean.csv` (+ optional Parquet)
- Quick aggregates: year, district, offense
- Reports:
  - `reports/data_dictionary.md`
  - `reports/missingness_before.csv`, `reports/missingness_after.csv`
  - `reports/data_quality_summary.md`

**Next (team):** Use clean dataset for 10 research questions × 2 visuals each (20 total).